syntax = "proto3";

import "google/protobuf/empty.proto";

package airmobisim;

service AirMobiSim{
    
    rpc Start (google.protobuf.Empty) returns (google.protobuf.Empty);

    rpc ExecuteOneTimeStep (google.protobuf.Empty) returns (ResponseQuery);

    rpc Finish (google.protobuf.Empty) returns (google.protobuf.Empty);

    rpc GetManagedHosts(google.protobuf.Empty) returns (UavList);

    rpc InsertWaypoints (WaypointList) returns (google.protobuf.Empty);

    rpc InsertUAV (StartUav) returns (google.protobuf.Empty);

    rpc DeleteUAV (Number) returns (google.protobuf.Empty);

    rpc getNumberCurrentUAV (google.protobuf.Empty) returns (Number);
    
    rpc SetDesiredSpeed (UavSetSpeed) returns (google.protobuf.Empty);

    rpc UpdateWaypoints (WaypointList) returns (google.protobuf.Empty);

    rpc DeleteWaypoint (WaypointList) returns (google.protobuf.Empty);          
}

message WaypointList{
    int32 id = 1;
    repeated Waypoint waypoint = 2;
}

message Waypoint
{
    int32 index = 1;
    double x = 2;
    double y = 3;
    double z = 4;
}


message UavList {
    repeated Uav uavs = 1;
}

message Uav {
    int32 id = 1;
    double x = 2;
    double y = 3;
    double z = 4;
    double speed = 5;
    double angle = 6;
}

message Response{
    int32 id = 1;
    double x = 2;
    double y = 3;
    double z = 4;
    double speed = 5;
    double angle = 6;
}

message ResponseQuery{
    repeated Response responses = 1;
}


message StartUav{
    int32 id = 1;
    repeated Coordinates coordinates = 2;
    double speed = 3;
    double angle = 4;
}

message Coordinates{
    double x = 1;
    double y = 2;
    double z = 3;
}


message Number{
    int32 num = 1;
}


message UavSetSpeed{
    int32 id = 1;
    double speed = 2;
}


